sharing/src/main/scala/io/delta/sharing/spark/DeltaSharingLogFileSystem.scala

```diff
@@ -353,6 +353,233 @@ private[sharing] object DeltaSharingLogFileSystem extends Logging {
+   *                           construct the delta log for the rpc no matter if there are files in
+   *                           that version or not, so DeltaSource can read delta actions from the
+   *                           starting version (instead from checkpoint).
+   * @param endingVersionOpt If set, used to construct the delta file (.json log file) until the
+   *                         given endingVersion. This is needed by DeltaSharingSource to construct
+   *                         the delta log for the rpc no matter if there are files in that version
+   *                         or not.
+   *                         NOTE: DeltaSource will not advance the offset if there are no files in
+   *                         a version of the delta log, but we still create the delta log file for
+   *                         that version to avoid missing delta log (json) files.
+   * @return ConstructedDeltaLogMetadata, which contains 3 fields:
+   *          - idToUrl: mapping from file id to pre-signed url
+   *          - minUrlExpirationTimestamp timestamp indicating the when to refresh pre-signed urls.
+   *            Both are used to register to CachedTableManager.
+   *          - maxVersion: the max version returned in the http response, used by
+   *            DeltaSharingSource to quickly understand the progress of rpcs from the server.
+   */
+  def constructLocalDeltaLogAcrossVersions(
+      lines: Seq[String],
+      customTablePath: String,
+      startingVersionOpt: Option[Long],
+      endingVersionOpt: Option[Long]): ConstructedDeltaLogMetadata = {
+    val startTime = System.currentTimeMillis()
+    assert(
+      startingVersionOpt.isDefined == endingVersionOpt.isDefined,
+      s"startingVersionOpt($startingVersionOpt) and endingVersionOpt($endingVersionOpt) should be" +
+      " both defined or not."
+    )
+    if (startingVersionOpt.isDefined) {
+      assert(
+        startingVersionOpt.get <= endingVersionOpt.get,
+        s"startingVersionOpt($startingVersionOpt) must be smaller than " +
+        s"endingVersionOpt($endingVersionOpt)."
+      )
+    }
+    var minVersion = Long.MaxValue
+    var maxVersion = 0L
+    var minUrlExpirationTimestamp: Option[Long] = None
+    val idToUrl = scala.collection.mutable.Map[String, String]()
+    val versionToDeltaSharingFileActions =
+      scala.collection.mutable.Map[Long, ArrayBuffer[model.DeltaSharingFileAction]]()
+    val versionToMetadata = scala.collection.mutable.Map[Long, model.DeltaSharingMetadata]()
+    val versionToJsonLogBuilderMap = scala.collection.mutable.Map[Long, ArrayBuffer[String]]()
+    val versionToJsonLogSize = scala.collection.mutable.Map[Long, Long]().withDefaultValue(0L)
+    var numFileActionsInMinVersion = 0
+    val versionToTimestampMap = scala.collection.mutable.Map[Long, Long]()
+    var startingMetadataLineOpt: Option[String] = None
+    var startingProtocolLineOpt: Option[String] = None
+
+    lines.foreach { line =>
```

Comment:
not now, but wondering if we can use the spark dataframe and delta actions to serializations to save the actions as checkpoint files? If yes, this code will be much smaller, but don't know the other implications.

Reply:
not for cdf I assume? because we want the delta library instead of the delta sharing code to produce the per version _change_type.

But for snapshot queries we can do that.